#                              **REFERENCES**
---------------------------------------------------------------------------------------------------------------------------------
Langmead B, Salzberg S. Fast gapped-read alignment with **Bowtie 2**. *Nature Methods.* 2012, 9:357-359.

Li H., Handsaker B., Wysoker A., Fennell T., Ruan J., Homer N., Marth G., Abecasis G., Durbin R. and 1000 Genome Project Data Processing Subgroup (2009). The Sequence alignment/map (SAM) format and SAMtools. Bioinformatics, 25, 2078-9.PMID: 19505943.

Yaari G, Bolen CR, Thakar J, Kleinstein SH (2013). Quantitative set analysis for gene expression: a method to quantify gene set differential expression including gene-gene correlations. Nucleic Acids Res. 2013 Oct;41(18):e170. doi: 10.1093/nar/gkt660. Epub Aug 5. PubMed PMID: 23921631; PubMed Central PMCID: PMC3794608.

https://support.illumina.com/content/dam/illumina-support/documents/documentation/software_documentation/bcl2fastq/bcl2fastq_letterbooklet_15038058brpmi.pdf

Andrews, S. **FastQC** A Quality Control tool for High Throughput Sequence Data. 2014 http://www.bioinformatics.babraham.ac.uk/projects/fastqc/. 

Stubbs T, Bonder M, Stark A, Krueger F. Multi-tissue DNA methylation age predictor in mouse. *Genome Biology*. 2017; 18:68. https://doi.org/10.1186/s13059-017-1203-5

Langmead B, Trapnell C, Pop M, Salzberg SL. Ultrafast and memory-efficient alignment of short DNA sequences to the human genome. *Genome Biology*.2009;10:R25. https://doi.org/10.1186/gb-2009-10-3-r25

Kim D, Pertea G, Trapnell C, Pimentel H, Kelley R, Salzberg SL. TopHat2: accurate alignment of transcriptomes in the presence of insertions, deletions and gene fusions.*Genome Biology*. 2013; 14(4):R36. doi: 10.1186/gb-2013-14-4-r36.

Trapnell C, Williams BA, Pertea G, Mortazavi AM, Kwan G, van Baren MJ, Salzberg SL, Wold B, Pachter L. Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation *Nature Biotechnology*. 2010; 28:511-515. http://dx.doi.org/10.1038/nbt.1621

Roberts A, Trapnell C, Donaghey J, Rinn JL, Pachter L. Improving RNA-Seq expression estimates by correcting for fragment bias. *Genome Biology*. 2011;14:R22. doi: 10.1186/gb-2011-12-3-r22.

Trapnell C, Hendrickson DG, Sauvageau M, Goff L, Rinn JL, Pachter L. Differential analysis of gene regulation at transcript resolution with RNA-seq. *Nature Biotechnology*. 2012;31(1):46-53.

Trapnell C, Roberts A, Goff L, Pertea G, Kim D, Kelley D, Pimentel H, Salzberg S, Rinn J, Pachter L. "Differential gene and transcript expression analysis of RNA-seq experiments with TopHat and Cufflinks" *Nature Protocols*. 2012; volume 7, pages 562–578; https://doi.org/10.1038/nprot.2012.016

Kim D, Langmead B, Salzberg SL (2015). HISAT: a fast spliced aligner with low memory requirements. Nat Methods. Apr;12(4):357-60. doi: 10.1038/nmeth.3317. Epub 2015 Mar 9. PubMed PMID: 25751142.

Pertea M, Pertea GM, Antonescu CM, Chang TC, Mendell JT, Salzberg SL (2015). StringTie enables improved reconstruction of a transcriptome from RNA-seq reads. Nat Biotechnol. Mar;33(3):290-5. doi: 10.1038/nbt.3122. Epub 2015 Feb 18. PubMed PMID: 25690850.
Frazee AC, Collado-Torres L, Jaffe AE and Leek JT. ballgown: Flexible, isoform-level differential expression analysis. R package version 2.0.0.

R Core Team. R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria. 2014; http://www.R-project.org/.

L. Goff, C. Trapnell and D. Kelley. "CummeRbund: Analysis, Exploration, Manipulation, and Visualization of Cufflinks High-Throughput Sequencing Data." 2013; R package version 2.20.0.

Liao Y, Smyth GK and Shi W (2014). featureCounts: an efficient general-purpose program for assigning sequence reads to genomic features. Bioinformatics, 30(7):923-30.

Harrow, J., Frankish, A., Gonzalez, J. M., Tapanari, E., Diekhans, M., Kokocinski, F., Hubbard, T. J. (2012). GENCODE: The reference human genome annotation for The ENCODE Project. Genome Research. doi:10.1101/gr.135350.111.

Robinson, M. D., McCarthy, D. J., Smyth, G. K. (2010). edgeR: a Bioconductor package for differential expression analysis of digital gene expression data. Bioinformatics (Oxford, England), 26, 139-140. doi:10.1093/bioinformatics/btp616.

Lun, ATL, Chen, Y, and Smyth, GK. It’s DE-licious: a recipe for differential expression analyses of RNA-seq experiments using quasi-likelihood methods in edgeR. *Methods in Molecular Biology*. 2016; 1418, 391–416

Smyth, G. K. Linear models and empirical Bayes methods for assessing differential expression in microarray experiments. *Statistical Applications in Genetics and Molecular Biology*. 2004; 3, No. 1, Article 3. http://www.statsci.org/smyth/pubs/ebayes.pdf

Ritchie, M. E., B. Phipson, D. Wu, Y. Hu, C. W. Law, W. Shi, and G. K. Smyth. “limma Powers Differential Expression Analyses for RNA-Sequencing and Microarray Studies.” *Nucleic Acids Research* 2015; 43 (7):e47.

Law, Charity W., Yunshun Chen, Wei Shi, and Gordon K. Smyth. “Voom: precision weights unlock linear model analysis tools for RNA-seq read counts.” *Genome Biology*. 2014; 15 (2). BioMed Central Ltd:R29+. https://doi.org/10.1186/gb-2014-15-2-r29.

Pertea, M., Kim, D., Pertea, G. M., Leek, J. T., & Salzberg, S. L. (2016). Transcript-level expression analysis of RNA-seq experiments with HISAT, StringTie and Ballgown. Nature protocols, 11(9), 1650.ISO 690

H. Fu J, Frazee AC, Collado-Torres L, Jaffe AE, Leek JT (2018). ballgown: Flexible, isoform-level differential expression analysis. R package version 2.12.0.

Li, B., Ruotti, V., Stewart, R. M., Thomson, J. A., & Dewey, C. N. (2009). RNA-Seq gene expression estimation with read mapping uncertainty. Bioinformatics, 26(4), 493-500.

McCarthy, J. D, Chen, Yunshun, Smyth, K. G (2012). “Differential expression analysis of multifactor RNA-Seq experiments with respect to biological variation.” Nucleic Acids Research, 40(10), 4288-4297.

Kim D, Langmead B and Salzberg SL. HISAT: a fast spliced aligner with low memory requirements. Nature Methods 2015

Pertea M, Pertea GM, Antonescu CM, Chang TC, Mendell JT & Salzberg SL. (2015).  "**StringTie** enables improved reconstruction of a transcriptome from RNA-seq reads." *Nature Biotechnology*, doi:10.1038/nbt.3122

